### PR TITLE
Provider metadata

### PR DESCRIPTION
**Pull request by tschmidtb51**
While drafting the prose for our last two roles:

- **CSAF lister**: lists URLs of known CSAF distribution points (where issuing parties (CSAF publisher, CSAF provider, CSAF trusted provider) publish their CSAF documents)
- **CSAF aggregator**:  lists URLs of known CSAF distribution points and mirrors the documents to provide an one-stop-shop for CSAF documents

I realized that the metadata which is needed to provide such lists is hard to collect. Especially, since we allow not only ROLIE but also a directory-based approach to distribute CSAF documents. It would be unclear to a user whether to expect the ROLIE or the  directory-based approach. Therefore, the metadata should be provided by the issuing party itself. As a result, I opened #289.

My description in #289 was a little bit brief. Therefore, I sketched this draft PR which should cover the points (some examples are still missing). I'll highlight the key points here:

- CSAF providers (including CSAF trusted providers) publish their own `provider-metadata.json`
- The `provider-metadata.json` can be found through one of the following ways (basically nothing changed here):
  - through `security.txt`
  - through Well-known URL
  - through DNS path
 - The `provider-metadata.json` lists whether the issuing party uses the ROLIE or the  directory-based approach. (Currently, only one is allowed at the same time, but we could change that for both being allowed.)
 - The `provider-metadata.json` provides the opportunity for issuing parties to prohibit that their CSAF documents are listed or mirrored at a CSAF aggregator.
 - The `provider-metadata.json` is a single file structure which is used for each role.
 - A CSAF aggregator can manually provide the metadata for a CSAF publisher and enable automation for that party which was not possible before.
- Add an entry to a list of known CSAF distribution points becomes pretty easy - just copy the `provider-metadata.json`
- ROLIE service documents become truly optional.

**Thoughts? Questions? Please comment!**

### CODE REVIEW

**Comment by tschmidtb51** [comment]
A rendered version is here: [7.1.7 Requirement 7: provider-metadata.json](https://github.com/tschmidtb51/csaf/blob/provider-metadata/csaf_2.0/prose/csaf-v2-editor-draft.md#717-requirement-7-provider-metadatajson) (unfortunately, not in diff view possible...).

**Review by sthagen — APPROVED** [review]
Nice. Thanks for the great work.

**Review by sthagen — APPROVED** [review]

**Review by santosomar — APPROVED** [review]

**Review by sthagen — APPROVED** [review]

**Review by mprpic — COMMENTED** [review]
What is the default value of `list_on_CSAF_aggregators` and `mirror_on_CSAF_aggregators` if a CSAF publisher does not provide a metadata JSON file?

Can we also add example `provider-metada.json` and `aggregator.json` files to https://github.com/oasis-tcs/csaf/tree/master/csaf_2.0/examples that could be validated against the respective schemas in a future CI test suite (minimal examples can still be present in prose as well of course)?

File: csaf_2.0/json_schema/provider_json_schema.json [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+          },
+          "url": {
+            "title": "URL of the key",
+            "description": "Contains the URL where the key can be retrieved.",
+            "$ref": "#/definitions/url_t"
+          }
+        }
+      }
+    },
+    "publisher": {
```
[the remark this answers is not in the document]
Comment (sthagen) [reply csaf_2.0/json_schema/provider_json_schema.json:196]: I am still not sure if our profiles are strictly additive (although I hope so).

File: csaf_2.0/json_schema/provider_json_schema.json [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+  "title": "CSAF provider metadata",
+  "description": "Representation of metadata information of a CSAF provider as a JSON document.",
+  "type": "object",
+  "definitions": {
+    "json_url_t": {
+      "title": "JSON URL type",
+      "description": "Contains a URL of a JSON file.",
+      "type": "string",
+      "format": "uri",
+      "pattern": "\\.json$"
```
[the remark this answers is not in the document]
Comment (sthagen) [reply csaf_2.0/json_schema/provider_json_schema.json:13]: I think the pattern is only anchored to the right, so It should match all files with a dot (`\`) followed by `json` and nothing else.

**Review by santosomar — APPROVED** [review]

File: csaf_2.0/json_schema/provider_json_schema.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (tschmidtb51) [reply csaf_2.0/json_schema/provider_json_schema.json:13]: Yes. That is the plan.

File: csaf_2.0/json_schema/provider_json_schema.json [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    },
+    "provider_url_t": {
+      "title": "Provider URL type",
+      "description": "Contains a URL of a provider-metadata.json.",
+      "type": "string",
+      "format": "uri",
+      "pattern": "/provider\\-metadata\\.json$"
```
[the remark this answers is not in the document]
Comment (tschmidtb51) [reply csaf_2.0/json_schema/provider_json_schema.json:20]: The `/` prefix should prevent that we get files named `example-company_provider-metadata.json`. The pattern is also only anchored to the right.

File: csaf_2.0/json_schema/provider_json_schema.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (tschmidtb51) [reply csaf_2.0/json_schema/provider_json_schema.json:196]: @sthagen This is unrelated to the profiles. (However, the `publisher` is part of the "Generic CSAF".)

@mprpic The purpose here is that an issuing party can decide what they would like to named in the list generated by the CSAF aggregator. I added a remark for that.

**Comment by tschmidtb51** [comment]
> What is the default value of `list_on_CSAF_aggregators` and `mirror_on_CSAF_aggregators` if a CSAF publisher does not provide a metadata JSON file?

Excellent question. I would suggest:
```
    "list_on_CSAF_aggregators": true,
    "mirror_on_CSAF_aggregators": false
```

Thoughts?

**Comment by tschmidtb51** [comment]
> Can we also add example `provider-metada.json` and `aggregator.json` files to https://github.com/oasis-tcs/csaf/tree/master/csaf_2.0/examples that could be validated against the respective schemas in a future CI test suite (minimal examples can still be present in prose as well of course)?

@mprpic  Yes. My plan included to resort the examples (into sub directories `csaf`, `provider-metadata`, `aggregator`, `rolie` and `directory`. Unfortunately, I hadn't had time to do that and patch the necessary files. (And to be honest: I wanted to wait until #296 is accepted to make sure we don't have to resolve merge conflicts in docker files.) But feel free to provide a PR...

**Comment by sthagen** [comment]
I propose to lift the current constraint forcing exclusivity in the statement (of this PR summary):
> - The `provider-metadata.json` lists whether the issuing party uses the ROLIE or the  directory-based approach. (Currently, only one is allowed at the same time, but we could change that for both being allowed.)

Rationale, how is a provider expected to offer directory based first and then ROLIE? Remove the existing directory infrastructure? I doubt this is a reasonable assumption.

Am I off track or is this a practical concern severe enough to lift the constraint?

File: csaf_2.0/json_schema/provider_json_schema.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (mprpic) [reply csaf_2.0/json_schema/provider_json_schema.json:196]: Perhaps I would note that explicitly then:

> The `publisher` object SHOULD match the one used in the CSAF documents of the issuing party **but can be set to whatever value a CSAF aggregator should display over any individual `publisher` values in the CSAF documents themselves.**

Or something like that...

**Comment by mprpic** [comment]
> I propose to lift the current constraint forcing exclusivity in the statement (of this PR summary):
> 
> > * The `provider-metadata.json` lists whether the issuing party uses the ROLIE or the  directory-based approach. (Currently, only one is allowed at the same time, but we could change that for both being allowed.)
> 
> Rationale, how is a provider expected to offer directory based first and then ROLIE? Remove the existing directory infrastructure? I doubt this is a reasonable assumption.
> 
> Am I off track or is this a practical concern severe enough to lift the constraint?

I would also consider removing this constraint and allow both approaches to exist.

File: csaf_2.0/json_schema/provider_json_schema.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (mprpic) [reply csaf_2.0/json_schema/provider_json_schema.json]: Ah, gotcha. LGTM!

**Review by sthagen — APPROVED** [review]

**Comment by tschmidtb51** [comment]
>> Am I off track or is this a practical concern severe enough to lift the constraint?
>
> I would also consider removing this constraint and allow both approaches to exist.

The contraint was removed in a3210a8.

> What is the default value of `list_on_CSAF_aggregators` and `mirror_on_CSAF_aggregators` if a CSAF publisher does not provide a metadata JSON file?

This is addressed in 76f1b70.

File: csaf_2.0/json_schema/provider_json_schema.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (tschmidtb51) [reply csaf_2.0/json_schema/provider_json_schema.json:195]: Thanks. I added that in 54eddb2.

**Review by sthagen — APPROVED** [review]

**Review by mprpic — APPROVED** [review]

File: csaf_2.0/prose/csaf-v2-editor-draft.md [hunk 4]
```diff
@@ -4736,7 +4736,7 @@ The party MUST provide a valid `provider-metadata.json` according to the schema
 
 **TODO: Point to a place where options and fields are described in details**
 
-*Examples 94: Minimal with ROLIE document*
+*Examples 94 Minimal with ROLIE document:*
```
Comment (tschmidtb51) [inline csaf_2.0/prose/csaf-v2-editor-draft.md:4739]: Thank you. I wanted to add that as a question but you were faster...

**Review by sthagen — APPROVED** [review]

**Review by sthagen — APPROVED** [review]

**Pull request merged**